clusterloader2/pkg/measurement/common/simple/wait_for_services.go

```diff
@@ -0,0 +1,147 @@
+)
+
+func init() {
+	measurement.Register(waitForServiceName, createWaitForServicesMeasurement)
+}
+
+func createWaitForServicesMeasurement() measurement.Measurement {
+	return &waitForServicesMeasurement{}
+}
+
+type waitForServicesMeasurement struct {
+	clientSet     clientset.Interface
+	namespace     string
+	labelSelector string
+	fieldSelector string
+	waitTimeout   time.Duration
+	store         *measurementutil.ServiceStore
+}
+
+// Execute waits until services are ready.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all-namespace scope is assumed.
+// "Start" action starts observation of the services, while "gather" waits until
+// desired number of services are ready.
+func (w *waitForServicesMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		w.clientSet = config.ClusterFramework.GetClientSets().GetClient()
+		w.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return nil, err
+		}
+		w.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		w.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		w.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultWaitForServicesTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, w.start()
+	case "gather":
+		desiredCount, err := util.GetInt(config.Params, "desiredCount")
+		if err != nil {
+			return nil, err
+		}
+		return nil, w.gather(desiredCount)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (w *waitForServicesMeasurement) start() error {
+	var err error
+	w.store, err = measurementutil.NewServiceStore(w.clientSet, w.namespace, w.labelSelector, w.fieldSelector)
+	if err != nil {
+		return fmt.Errorf("service store creation error: %v", err)
+	}
+	return nil
+}
+
+func (w *waitForServicesMeasurement) gather(desiredCount int) error {
+	if w.store == nil {
+		klog.Errorf("%s: measurement is not started", w)
+		return fmt.Errorf("no service store")
+	}
+
+	created, available := 0, 0
+	selectorsString := measurementutil.CreateSelectorsString(w.namespace, w.labelSelector, w.fieldSelector)
+	for {
+		select {
+		case <-time.After(w.waitTimeout):
+			return fmt.Errorf("timeout while waiting for %d services to be available in namespace '%v' with labels '%v' and fields '%v' - only %d found running",
+				desiredCount, w.namespace, w.labelSelector, w.fieldSelector, available)
+		case <-time.After(defaultWaitForServicesInterval):
+			services := w.store.List()
+			created = len(services)
+			available = 0
+			// TODO: Measurement waits only for load balancers.
```

Comment:
Can we add another LoadBalancer service specific measurement that will run "resp, err := http.Get("http://<LoadBalancerIP>:servicePort")" and check that err is nil and resp is valid?

Reply:
You mean probe that service?

There is a separate work trying to start probes and I would like to have those consistent.

But my main question is: is that probe going to measure:
(a) whether the service is healthy over longer period of time?
or
(b) ping the service as long as it doesn't respond to get the latency of when we get the first successful response

Reply:
Option b). Goal is to measure the latency from issuing the service creation command to having the load balancer respond.